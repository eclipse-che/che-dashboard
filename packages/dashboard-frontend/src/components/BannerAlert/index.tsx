/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';

import BannerAlertBranding from '@/components/BannerAlert/Branding';
import BannerAlertCustomWarning from '@/components/BannerAlert/Custom';
import BannerAlertNoNodeAvailable from '@/components/BannerAlert/NoNodeAvailable';
import BannerAlertWebSocket from '@/components/BannerAlert/WebSocket';

type Props = unknown;

type State = {
  bannerAlerts: React.ReactElement[];
};

export class BannerAlert extends React.PureComponent<Props, State> {
  constructor(props) {
    super(props);
    this.state = {
      bannerAlerts: [
        <BannerAlertWebSocket key="BannerAlertWebSocket"></BannerAlertWebSocket>,
        <BannerAlertBranding key="BannerAlertBranding"></BannerAlertBranding>,
        <BannerAlertCustomWarning key="BannerAlertCustomWarning"></BannerAlertCustomWarning>,
        <BannerAlertNoNodeAvailable key="BannerAlertNoNodeAvailable"></BannerAlertNoNodeAvailable>,
      ],
    };
  }

  render() {
    const banners = this.state.bannerAlerts;
    return (
      <div>
        {banners.map(banner => (
          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
          <div key={banner.key!}>{banner}</div>
        ))}
      </div>
    );
  }
}
