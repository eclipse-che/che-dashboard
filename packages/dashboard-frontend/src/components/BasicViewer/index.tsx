/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { defaultHighlightStyle, syntaxHighlighting } from '@codemirror/language';
import { EditorState } from '@codemirror/state';
import { EditorView } from '@codemirror/view';
import React from 'react';

import styles from '@/components/BasicViewer/index.module.css';

export type Props = {
  value: string;
  id: string;
};

export class BasicViewer extends React.PureComponent<Props> {
  private editor: EditorView | undefined;

  public componentDidMount(): void {
    const parent = window.document.querySelector(`#${this.props.id}`);
    if (parent) {
      const state = this.getEditorState();
      this.editor = new EditorView({
        state,
        parent,
      });
    }
  }

  public componentWillUnmount(): void {
    if (this.editor) {
      this.editor.destroy();
    }
  }

  componentDidUpdate(prevProps: Readonly<Props>): void {
    if (this.editor && this.props.value !== prevProps.value) {
      const state = this.getEditorState();
      this.editor.setState(state);
    }
  }

  private getEditorState(): EditorState {
    return EditorState.create({
      doc: this.props.value,
      extensions: [syntaxHighlighting(defaultHighlightStyle), EditorState.readOnly.of(true)],
    });
  }

  public render(): React.ReactElement {
    return (
      <div className={styles.basicViewer}>
        <div id={this.props.id}></div>
      </div>
    );
  }
}
