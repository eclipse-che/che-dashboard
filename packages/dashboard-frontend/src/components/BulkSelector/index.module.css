/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

.selector .pf-c-select__menu {
  overflow-y: auto;
  min-width: 200px;
  max-height: 310px;
}
