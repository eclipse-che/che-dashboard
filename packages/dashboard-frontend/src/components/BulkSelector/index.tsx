/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { Select, SelectOption, SelectOptionObject, SelectVariant } from '@patternfly/react-core';
import React from 'react';

import styles from '@/components/BulkSelector/index.module.css';

export type Props = {
  onChange: (tags: string[]) => void;
  list: string[];
  placeholderText: string;
};

export type State = {
  isOpen: boolean;
  selected: string[];
  options: React.ReactElement[];
};

export class BulkSelector extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      isOpen: false,
      selected: [],
      options: [],
    };
  }

  public componentDidMount(): void {
    const options = this.getOptions(this.props.list);
    this.setState({
      options,
    });
  }

  public componentDidUpdate(prevProps: Props): void {
    const { list } = this.props;
    if (prevProps.list !== list) {
      const options = this.getOptions(list);
      this.setState({
        options,
      });
    }
  }

  private getOptions(list: string[]): React.ReactElement[] {
    return list.map(tag => <SelectOption key={tag} value={tag} />);
  }

  private onToggle(isOpen: boolean): void {
    this.setState({
      isOpen,
    });
  }

  private onSelect(selection: string) {
    const selected = [...this.state.selected];
    const index = selected.indexOf(selection);
    if (index > -1) {
      selected.splice(index, 1);
    } else {
      selected.push(selection);
    }
    this.setState({ selected });
    this.props.onChange(selected);
  }

  render(): React.ReactNode {
    const { placeholderText } = this.props;
    const { isOpen, selected, options } = this.state;

    if (options.length === 0) {
      return <></>;
    }

    return (
      <Select
        className={styles.selector}
        variant={SelectVariant.checkbox}
        onToggle={isOpen => this.onToggle(isOpen)}
        onSelect={(
          _event: React.MouseEvent | React.ChangeEvent,
          selection: string | SelectOptionObject,
        ) => this.onSelect(selection.toString())}
        selections={selected}
        isOpen={isOpen}
        placeholderText={placeholderText}
        isGrouped
      >
        {options}
      </Select>
    );
  }
}
