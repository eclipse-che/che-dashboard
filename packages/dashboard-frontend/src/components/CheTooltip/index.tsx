/*
 * Copyright (c) 2018-2023 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { Tooltip, TooltipPosition } from '@patternfly/react-core';
import React from 'react';

type Props = {
  children: React.ReactElement;
  content: React.ReactNode;
  position?: TooltipPosition;
};

class CheTooltip extends React.PureComponent<Props> {
  public render(): React.ReactElement {
    const { content, position, children } = this.props;

    return (
      <Tooltip
        exitDelay={3000}
        isContentLeftAligned={true}
        position={position ? position : TooltipPosition.right}
        content={content}
        style={{ border: '1px solid', borderRadius: '3px', opacity: '0.9' }}
      >
        {children}
      </Tooltip>
    );
  }
}

export default CheTooltip;
