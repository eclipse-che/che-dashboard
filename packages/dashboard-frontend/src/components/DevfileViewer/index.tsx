/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import 'codemirror/mode/yaml/yaml';
import 'codemirror/lib/codemirror.css';
import '@/components/DevfileViewer/theme/eclipse-che.css';

import { fromTextArea as CMEditor } from 'codemirror';
import React from 'react';

import styles from '@/components/DevfileViewer/index.module.css';

export type Props = {
  isActive: boolean;
  isExpanded: boolean;
  value: string;
  id: string;
};

export class DevfileViewer extends React.PureComponent<Props> {
  private editor: CMEditor | undefined;

  constructor(props: Props) {
    super(props);
  }

  public componentDidMount(): void {
    const element = window.document.querySelector(
      `.${styles.devfileViewer} textarea#${this.props.id}`,
    );
    if (element !== null) {
      const editor = new CMEditor(element, {
        mode: 'yaml',
        theme: 'eclipse-che',
        lineNumbers: true,
        lineWrapping: true,
        readOnly: true,
        autoRefresh: true,
      });
      editor.setSize(`100%`, `100%`);
      editor.setValue(this.props.value);

      this.editor = editor;
    }
  }

  public componentDidUpdate(): void {
    if (!this.editor) {
      return;
    }

    this.editor.setValue(this.props.value);
    this.editor.refresh();
    this.editor.focus();
  }

  public render(): React.ReactElement {
    return (
      <div className={styles.devfileViewer}>
        <textarea id={this.props.id}></textarea>
      </div>
    );
  }
}
