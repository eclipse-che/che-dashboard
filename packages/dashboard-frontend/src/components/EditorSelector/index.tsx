/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import {
  Accordion,
  AccordionContent,
  AccordionItem,
  AccordionToggle,
  Button,
  Flex,
  FlexItem,
  Panel,
  PanelHeader,
  PanelMain,
  PanelMainBody,
  Title,
} from '@patternfly/react-core';
import React from 'react';
import { connect, ConnectedProps } from 'react-redux';

import { EditorDefinition } from '@/components/EditorSelector/Definition';
import { EditorGallery } from '@/components/EditorSelector/Gallery';
import { AppState } from '@/store';
import { selectEditors } from '@/store/Plugins/chePlugins/selectors';

const DOCS_DEFINING_A_COMMON_IDE =
  'https://eclipse.dev/che/docs/stable/end-user-guide/defining-a-common-ide/';

type AccordionId = 'selector' | 'definition';

export type Props = MappedProps & {
  defaultEditorId: string;
  onSelect: (editorDefinition: string | undefined, editorImage: string | undefined) => void;
};
export type State = {
  definitionEditorValue: string | undefined;
  definitionImageValue: string | undefined;

  selectorEditorValue: string;

  expandedId: AccordionId | undefined;
};

class EditorSelector extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      definitionEditorValue: undefined,
      definitionImageValue: undefined,

      selectorEditorValue: '',

      expandedId: 'selector',
    };
  }

  private handleSelectorValueChange(editorId: string): void {
    this.setState({ selectorEditorValue: editorId });
    this.props.onSelect(editorId, undefined);
  }

  private handleDefinitionValueChange(
    editorDefinition: string | undefined,
    editorImage: string | undefined,
  ): void {
    this.setState({
      definitionEditorValue: editorDefinition,
      definitionImageValue: editorImage,
    });
    this.props.onSelect(editorDefinition, editorImage);
  }

  private handleToggle(expandedId: AccordionId): void {
    const { onSelect } = this.props;

    this.setState({
      expandedId: this.state.expandedId === expandedId ? this.state.expandedId : expandedId,
    });

    const { definitionEditorValue, definitionImageValue, selectorEditorValue } = this.state;

    if (expandedId === 'selector') {
      onSelect(selectorEditorValue, undefined);
    } else {
      onSelect(definitionEditorValue, definitionImageValue);
    }
  }

  render(): React.ReactElement {
    const { defaultEditorId, editors } = this.props;
    const { definitionEditorValue, definitionImageValue, selectorEditorValue, expandedId } =
      this.state;

    return (
      <Panel>
        <PanelHeader>
          <Title headingLevel="h3">Editor Selector</Title>
        </PanelHeader>
        <PanelMain>
          <PanelMainBody>
            <Accordion asDefinitionList={false}>
              <AccordionItem>
                <AccordionToggle
                  onClick={() => {
                    this.handleToggle('selector');
                  }}
                  isExpanded={expandedId === 'selector'}
                  id="accordion-item-selector"
                >
                  Choose an Editor
                </AccordionToggle>

                <AccordionContent
                  isHidden={expandedId !== 'selector'}
                  data-testid="editor-gallery-content"
                >
                  <Panel>
                    <PanelMain>
                      <PanelMainBody>
                        <EditorGallery
                          defaultEditorId={defaultEditorId}
                          editors={editors}
                          selectedEditorId={selectorEditorValue}
                          onSelect={editorId => this.handleSelectorValueChange(editorId)}
                        />
                      </PanelMainBody>
                    </PanelMain>
                  </Panel>
                </AccordionContent>
              </AccordionItem>

              <AccordionItem>
                <AccordionToggle
                  onClick={() => {
                    this.handleToggle('definition');
                  }}
                  isExpanded={expandedId === 'definition'}
                  id="accordion-item-definition"
                >
                  Use an Editor Definition
                </AccordionToggle>

                <AccordionContent
                  isHidden={expandedId !== 'definition'}
                  data-testid="editor-definition-content"
                >
                  <Panel>
                    <PanelMain>
                      <PanelMainBody>
                        <EditorDefinition
                          editorDefinition={definitionEditorValue}
                          editorImage={definitionImageValue}
                          onChange={(editorDefinition, editorImage) =>
                            this.handleDefinitionValueChange(editorDefinition, editorImage)
                          }
                        />
                      </PanelMainBody>
                    </PanelMain>
                  </Panel>
                </AccordionContent>
              </AccordionItem>
            </Accordion>
            <Flex>
              <FlexItem align={{ default: 'alignRight' }}>
                <Button
                  component="a"
                  href={DOCS_DEFINING_A_COMMON_IDE}
                  variant="link"
                  isInline
                  target="_blank"
                >
                  How to specify and use a custom editor
                </Button>
              </FlexItem>
            </Flex>
          </PanelMainBody>
        </PanelMain>
      </Panel>
    );
  }
}

const mapStateToProps = (state: AppState) => ({
  editors: selectEditors(state),
});

const connector = connect(mapStateToProps, null, null, {
  // forwardRef is mandatory for using `@react-mock/state` in unit tests
  forwardRef: true,
});

type MappedProps = ConnectedProps<typeof connector>;
export default connector(EditorSelector);
