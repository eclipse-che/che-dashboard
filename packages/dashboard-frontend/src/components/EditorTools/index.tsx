/*
 * Copyright (c) 2018-2023 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { ApplicationId, helpers } from '@eclipse-che/common';
import { AlertVariant, Button, Divider, Flex, FlexItem } from '@patternfly/react-core';
import {
  CompressIcon,
  CopyIcon,
  DownloadIcon,
  ExpandIcon,
  ExternalLinkSquareAltIcon,
} from '@patternfly/react-icons';
import React from 'react';
import CopyToClipboard from 'react-copy-to-clipboard';
import { ConnectedProps, connect } from 'react-redux';
import { ToggleBarsContext } from '../../contexts/ToggleBars';
import { lazyInject } from '../../inversify.config';
import { AppAlerts } from '../../services/alerts/appAlerts';
import devfileApi, { isDevWorkspace, isDevfileV2 } from '../../services/devfileApi';
import stringify from '../../services/helpers/editor';
import { AlertItem } from '../../services/helpers/types';
import { AppState } from '../../store';
import { actionCreators } from '../../store/BannerAlert';
import { selectApplications } from '../../store/ClusterInfo/selectors';
import styles from './index.module.css';
import { WorkspaceAdapter } from '../../services/workspace-adapter';

type Props = MappedProps & {
  devfileOrDevWorkspace: devfileApi.DevWorkspace | devfileApi.Devfile;
  handleExpand: (isExpand: boolean) => void;
};

type State = {
  copied?: boolean;
  isExpanded: boolean;
  contentText: string;
  contentBlobUrl: string;
};

class EditorTools extends React.PureComponent<Props, State> {
  static contextType = ToggleBarsContext;
  readonly context: React.ContextType<typeof ToggleBarsContext>;

  private copiedTimer: number | undefined;

  @lazyInject(AppAlerts)
  private readonly appAlerts: AppAlerts;

  constructor(props: Props) {
    super(props);

    this.state = {
      isExpanded: false,
      contentText: '',
      contentBlobUrl: '',
    };
  }

  public componentDidMount(): void {
    this.init();
  }

  public componentDidUpdate(): void {
    this.init();
  }

  private init() {
    const { devfileOrDevWorkspace } = this.props;
    try {
      const contentText = stringify(devfileOrDevWorkspace);
      if (contentText !== this.state.contentText) {
        const contentBlobUrl = URL.createObjectURL(
          new Blob([contentText], { type: 'application/x-yaml' }),
        );
        this.setState({ contentText, contentBlobUrl });
      }
    } catch (e) {
      this.showAlert({
        key: 'editor-tools-create blob-url-fails',
        variant: AlertVariant.danger,
        title: helpers.errors.getMessage(e),
      });
    }
  }

  private showAlert(alert: AlertItem): void {
    this.appAlerts.showAlert(alert);
  }

  private handleExpand(): void {
    if (this.state.isExpanded) {
      this.context.showAll();
      const isExpanded = false;
      this.setState({ isExpanded });
      this.props.handleExpand(isExpanded);
    } else {
      this.context.hideAll();
      const isExpanded = true;
      this.setState({ isExpanded });
      this.props.handleExpand(isExpanded);
    }
  }

  private onCopyToClipboard(): void {
    this.setState({ copied: true });
    if (this.copiedTimer) {
      clearTimeout(this.copiedTimer);
    }
    this.copiedTimer = window.setTimeout(() => {
      this.setState({ copied: false });
    }, 3000);
  }

  private buildOpenShiftConsoleItem(): React.ReactElement | undefined {
    const { applications, devfileOrDevWorkspace } = this.props;
    const clusterConsole = applications.find(app => app.id === ApplicationId.CLUSTER_CONSOLE);

    if (isDevfileV2(devfileOrDevWorkspace)) {
      return;
    }
    const devWorkspace = devfileOrDevWorkspace;

    if (!clusterConsole) {
      return;
    }

    const devWorkspaceOpenShiftConsoleUrl = WorkspaceAdapter.buildClusterConsoleUrl(
      devWorkspace,
      clusterConsole.url,
    );

    return (
      <>
        <FlexItem>
          <Button
            className={styles.button}
            component="a"
            variant="link"
            href={devWorkspaceOpenShiftConsoleUrl}
            target="_blank"
            icon={<ExternalLinkSquareAltIcon />}
          >
            {clusterConsole.title}
          </Button>
        </FlexItem>
        <Divider isVertical />
      </>
    );
  }

  public render(): React.ReactElement {
    const { devfileOrDevWorkspace } = this.props;
    const { contentText, contentBlobUrl, isExpanded, copied } = this.state;

    const { name } = devfileOrDevWorkspace.metadata;

    const openshiftConsoleItem = this.buildOpenShiftConsoleItem();

    return (
      <div className={styles.editorTools}>
        <Flex>
          {openshiftConsoleItem}
          <FlexItem>
            <CopyToClipboard text={contentText} onCopy={() => this.onCopyToClipboard()}>
              <Button variant="link" className={styles.button}>
                <CopyIcon />
                {copied ? 'Copied' : 'Copy to clipboard'}
              </Button>
            </CopyToClipboard>
          </FlexItem>
          <Divider isVertical />
          <FlexItem>
            <a
              className={styles.button}
              download={`${name}.${
                isDevWorkspace(devfileOrDevWorkspace) ? 'workspace' : 'devfile'
              }.yaml`}
              href={contentBlobUrl}
            >
              <DownloadIcon />
              Download
            </a>
          </FlexItem>
          <Divider isVertical />
          <FlexItem>
            <Button className={styles.button} variant="link" onClick={() => this.handleExpand()}>
              {isExpanded ? (
                <React.Fragment>
                  <CompressIcon />
                  Compress
                </React.Fragment>
              ) : (
                <React.Fragment>
                  <ExpandIcon />
                  Expand
                </React.Fragment>
              )}
            </Button>
          </FlexItem>
        </Flex>
      </div>
    );
  }
}

const mapStateToProps = (state: AppState) => ({
  applications: selectApplications(state),
});

const connector = connect(mapStateToProps, actionCreators);

type MappedProps = ConnectedProps<typeof connector>;

export default connector(EditorTools);
