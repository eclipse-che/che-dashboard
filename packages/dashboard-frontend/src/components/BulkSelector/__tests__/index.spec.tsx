/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import userEvent, { UserEvent } from '@testing-library/user-event';
import React from 'react';

import { BulkSelector } from '@/components/BulkSelector';
import getComponentRenderer, { screen } from '@/services/__mocks__/getComponentRenderer';

const { createSnapshot, renderComponent } = getComponentRenderer(getComponent);

const mockOnChange = jest.fn();

describe('BulkSelector', () => {
  let list: string[];
  let user: UserEvent;

  beforeEach(() => {
    list = ['tag1', 'tag2', 'tag3'];
    user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
  });

  afterEach(() => {
    jest.clearAllMocks();
    mockOnChange.mockReset();
  });

  describe('kebab toggle', () => {
    test('snapshot with empty list', () => {
      const snapshot = createSnapshot([]);
      expect(snapshot.toJSON()).toMatchSnapshot();
    });

    test('snapshot with elements', () => {
      const snapshot = createSnapshot(list, 'Bulk Selector');
      expect(snapshot.toJSON()).toMatchSnapshot();
    });

    test('open dropdown', async () => {
      const { reRenderComponent } = renderComponent(list);
      const toggle = screen.queryByRole('button');

      expect(toggle).toBeTruthy();

      // dropdown menu is not visible
      expect(toggle).toHaveAttribute('aria-expanded', 'false');
      list.forEach(tag => {
        expect(screen.queryByText(tag)).toBeFalsy();
      });

      // toggle dropdown
      await user.click(toggle!);

      // now the dropdown menu is visible
      expect(toggle).toHaveAttribute('aria-expanded', 'true');
      list.forEach(tag => {
        expect(screen.queryByText(tag)).toBeTruthy();
      });
      // update menu items
      list = ['newTag1', 'newTag2'];
      reRenderComponent(list);
      list.forEach(tag => {
        expect(screen.queryByText(tag)).toBeTruthy();
      });
    });
  });

  test('handling selection', async () => {
    renderComponent(list);
    // toggle dropdown
    await user.click(screen.queryByRole('button')!);
    // select the first item
    await user.click(screen.queryByText(list[0])!);
    // expect selected item to be in the selected state
    expect(mockOnChange).toHaveBeenCalledWith([list[0]]);
    // select the second item
    await user.click(screen.queryByText(list[1])!);
    // expect selected item to be in the selected state
    expect(mockOnChange).toHaveBeenCalledWith([list[0], list[1]]);
    // deselect the first item
    await user.click(screen.queryByText(list[0])!);
    // expect selected item to be in the selected state
    expect(mockOnChange).toHaveBeenCalledWith([list[1]]);
  });
});

function getComponent(
  list: string[],
  placeholderText: string = '-- Select --',
): React.ReactElement {
  const component = (
    <BulkSelector list={list} placeholderText={placeholderText} onChange={mockOnChange} />
  );

  return component;
}
