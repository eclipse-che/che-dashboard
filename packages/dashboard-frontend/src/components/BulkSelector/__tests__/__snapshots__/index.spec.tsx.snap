// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BulkSelector kebab toggle snapshot with elements 1`] = `
<div
  className="pf-c-select selector"
  data-ouia-component-id="OUIA-Generated-Select-checkbox-1"
  data-ouia-component-type="PF4/Select"
  data-ouia-safe={true}
>
  <button
    aria-expanded={false}
    aria-haspopup={null}
    aria-label="Options menu"
    aria-labelledby=" pf-select-toggle-id-0"
    className="pf-c-select__toggle"
    disabled={false}
    id="pf-select-toggle-id-0"
    onClick={[Function]}
    onKeyDown={[Function]}
    type="button"
  >
    <div
      className="pf-c-select__toggle-wrapper"
    >
      <span
        className="pf-c-select__toggle-text"
      >
        Bulk Selector
      </span>
    </div>
    <span
      className="pf-c-select__toggle-arrow"
    >
      <svg
        aria-hidden={true}
        aria-labelledby={null}
        fill="currentColor"
        height="1em"
        role="img"
        style={
          {
            "verticalAlign": "-0.125em",
          }
        }
        viewBox="0 0 320 512"
        width="1em"
      >
        <path
          d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
        />
      </svg>
    </span>
  </button>
</div>
`;

exports[`BulkSelector kebab toggle snapshot with empty list 1`] = `null`;
